File: problems/37-bitmask-dp/001-partition-to-k-equal-sum-subsets/java/Solution.java
import java.util.Arrays;

public class Solution {
    private int[] nums;
    private int target;
    private Boolean[] memo;

    public boolean canPartitionKSubsets(int[] nums, int k) {
        int total = 0;
        for (int value : nums) total += value;
        if (k <= 0 || total % k != 0) return false;

        this.nums = nums.clone();
        this.target = total / k;
        Arrays.sort(this.nums);
        int n = nums.length;
        memo = new Boolean[1 << n];
        return dfs(0, 0);
    }

    private boolean dfs(int mask, int currentSum) {
        if (mask == (1 << nums.length) - 1) return currentSum == 0;
        if (memo[mask] != null) return memo[mask];

        for (int i = nums.length - 1; i >= 0; i--) {
            if ((mask & (1 << i)) != 0) continue;
            if (currentSum + nums[i] > target) continue;

            int nextSum = currentSum + nums[i];
            int nextMask = mask | (1 << i);
            if (nextSum == target) nextSum = 0;

            if (dfs(nextMask, nextSum)) return memo[mask] = true;
            if (currentSum == 0) break;
            while (i > 0 && nums[i] == nums[i - 1]) i--;
        }
        return memo[mask] = false;
    }
}

File: problems/37-bitmask-dp/001-partition-to-k-equal-sum-subsets/python/solution.py
from functools import lru_cache


def can_partition_k_subsets(nums: list[int], k: int) -> bool:
    total = sum(nums)
    if k <= 0 or total % k != 0:
        return False

    nums = sorted(nums, reverse=True)
    target = total // k
    n = len(nums)

    @lru_cache(maxsize=None)
    def dfs(mask: int, current_sum: int) -> bool:
        if mask == (1 << n) - 1:
            return current_sum == 0

        for i, value in enumerate(nums):
            if mask & (1 << i):
                continue
            if current_sum + value > target:
                continue

            next_sum = current_sum + value
            if next_sum == target:
                next_sum = 0

            if dfs(mask | (1 << i), next_sum):
                return True

            if current_sum == 0:
                break

        return False

    return dfs(0, 0)
